pkg/referrerstore/oras/authprovider/k8secret_authprovider.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authprovider
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/docker/cli/cli/config"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type k8SecretProviderFactory struct{}
+type k8SecretAuthProvider struct {
+	secrets []*core.Secret
+}
+
+type secretConfig struct {
+	SecretName string `json:"secretName"`
+	Namespace  string `json:"namespace,omitempty"`
+}
+
+type k8SecretAuthProviderConf struct {
+	Name               string         `json:"name"`
+	ServiceAccountName string         `json:"serviceAccountName,omitempty"`
+	Secrets            []secretConfig `json:"secrets,omitempty"`
+}
+
+const defaultName = "default"
+const ratifyNamespaceEnv = "RATIFY_NAMESPACE"
+
+// init calls Register for our k8s-secrets provider
+func init() {
+	Register("k8s-secrets", &k8SecretProviderFactory{})
+}
+
+// Create returns a k8AuthProvider instance after parsing auth config and resolving
+// named K8 secrets
+func (s *k8SecretProviderFactory) Create(authProviderConfig AuthProviderConfig) (AuthProvider, error) {
+
+	conf := k8SecretAuthProviderConf{}
+	authProviderConfigBytes, err := json.Marshal(authProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(authProviderConfigBytes, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse auth provider configuration: %v", err)
+	}
+
+	clusterConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clusterClientSet, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if conf.ServiceAccountName == "" {
+		conf.ServiceAccountName = defaultName
+	}
+
+	// get name of namespace ratify is running in
+	ratifyNamespace := os.Getenv(ratifyNamespaceEnv)
+	if ratifyNamespace == "" {
+		return nil, fmt.Errorf("environment variable %s not set", ratifyNamespaceEnv)
+	}
+
+	var k8secrets []*core.Secret
+	ctx := context.Background()
+
+	// iterate through config secrets, resolve each secret, and store in map
```

Comment:
If the secret value is updated, does Ratify needs restart to pick the new values?

Reply:
Yes, this is definitely an issue. We could pull the secrets from every time the Provide function is called, but I believe we are going to start running into timeout issues. Also, a potentially larger issue is that we are currently caching the ORAS registry client objects so that we don't have to call Provide every time we want to make a registry client. However, this can cause issues when the K8 secret might have been updated after the registry client was created.

Reply:
Updated so it now pulls the secrets from the cluster on every call to Provide. However, we still cache the registry client so that we don't have to call Provide every time. I think maybe we should add an expiry of 24 hours for the K8 secret credential so that the cache is invalidated and Ratify is forced to pull again.

Reply:
Could we add logic to refresh cred if artifact pull fails? what is our retry logic if pull fails?

Reply:
Another approach might be to ping the registry server to see if auth succeeds and no errors are returned upon a registry client cache hit. ORAS Go v2 exposes a [Ping function](https://github.com/oras-project/oras-go/blob/4a2d5071dad3e30d98000c8620fe9b8573c3d739/registry/remote/registry.go#L81). Although this would add another request so we'd have to measure how much latency this would add.

Reply:
Discussed with @akashsinghal , we can merge changeset now and log an issue to discuss the cache refresh strategy.  @akashsinghal  please open an item with the options discussed.